File: Data Structures & Algorithms/coin-change/submission-1.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[][] memo = new int[n][amount + 1];
        for (int i = 0; i < n; i++) {
            Arrays.fill(memo[i], -1);
        }

        int result = coinChangeHelper(0, coins, amount, memo);
        return result == Integer.MAX_VALUE ? -1 : result;
    }

    private int coinChangeHelper(int i, int[] coins, int amount, int[][] memo) {
        if (i >= coins.length || amount < 0) {
            return Integer.MAX_VALUE;
        }
        if (amount == 0) {
            return 0;
        } 

        if (memo[i][amount] != -1) {
            return memo[i][amount];
        }

        int skip = coinChangeHelper(i + 1, coins, amount, memo);

        int take = Integer.MAX_VALUE;
        if (coins[i] <= amount) {
            int result = coinChangeHelper(i, coins, amount - coins[i], memo);
            if (result != Integer.MAX_VALUE) {
                take = 1 + result;
            }
        }

        return memo[i][amount] = Math.min(take, skip);
    }
}

File: Data Structures & Algorithms/coin-change-ii/submission-1.go
func change(amount int, coins []int) int {
	n := len(coins)

	memo := make([][]int, n)
	for i := range memo {
		memo[i] = make([]int, amount + 1)

		for j := range memo[i] {
			memo[i][j] = -1
		}
	}

    return coinChangeHelper(0, amount, coins, memo)
}

func coinChangeHelper(i int, amount int, coins []int, memo [][]int) int{
	// base case
	if amount == 0 {
		return 1
	}

	if i >= len(coins) || amount < 0 {
		return 0
	}

	if memo[i][amount] != -1 {
		return memo[i][amount]
	}

	var take = 0
	if coins[i] <= amount {
		take = coinChangeHelper(i, amount - coins[i], coins, memo)
	}

	skip := coinChangeHelper(i + 1, amount, coins, memo)

	memo[i][amount] = take + skip
	return take + skip
}
